pkg/gitutil/gitutil.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2018 clair authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gitutil implements an easy way to update a git repository to a local
+// temporary directory.
+package gitutil
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ErrFailedClone is returned when a git clone is unsuccessful.
+var ErrFailedClone = errors.New("failed to clone git repository")
+
+// ErrFailedRevParse is returned when a git rev-parse is unsuccessful.
+var ErrFailedRevParse = errors.New("failed to rev-parse git repository")
+
+// ErrFailedPull is returned when a git pull is unsuccessful.
+var ErrFailedPull = errors.New("failed to pull git repository")
+
+// pull performs a git pull on the provided path and returns the commit SHA
+// for the HEAD reference.
+func pull(path string) (head string, err error) {
+	// Prepare a command to pull the repository.
+	cmd := exec.Command("git", "pull")
+	cmd.Dir = path
+
+	// Execute the command.
+	var commandOutput []byte
+	commandOutput, err = cmd.CombinedOutput()
+	if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"path":   path,
+			"output": string(commandOutput),
+		}).Error("failed to git rev-parse repository")
+		err = ErrFailedPull
+		return
+	}
+
+	return revParseHead(path)
+}
+
+// CloneOrPull performs a git pull if the desiredPath exists on the filesystem.
+// Otherwise, it generates a new temporary directory and performs a git clone
+// to that path.
```

Comment:
Different from the documentation here, the code only generate a new temporary directory when the desired path is not given. Is that the expected behavior?